diff --git a/diablo.cpp b/diablo.cpp
index 6d53647..54e7a7a 100644
--- a/diablo.cpp
+++ b/diablo.cpp
@@ -359,9 +359,9 @@ static void print_help_and_exit()
 	printf("    %-20s %-30s\n", "-n", "Skip startup videos");
 	printf("    %-20s %-30s\n", "-f", "Display frames per second");
 	printf("    %-20s %-30s\n", "-x", "Run in windowed mode");
+	printf("    %-20s %-30s\n", "-d", "Increaased item drops");
 #ifdef _DEBUG
 	printf("\nDebug options:\n");
-	printf("    %-20s %-30s\n", "-d", "Increaased item drops");
 	printf("    %-20s %-30s\n", "-w", "Enable cheats");
 	printf("    %-20s %-30s\n", "-$", "Enable god mode");
 	printf("    %-20s %-30s\n", "-^", "Enable god mode and debug tools");
@@ -389,6 +389,8 @@ void diablo_parse_flags(int argc, char **argv)
 			diablo_quit(0);
 		} else if (strcasecmp("--data-dir", argv[i]) == 0) {
 			basePath = argv[++i];
+		} else if (strcasecmp("-d", argv[i]) == 0) {
+			debug_mode_key_d = TRUE;
 #ifdef _WIN32
 			if (basePath.back() != '\\')
 				basePath += '\\';
@@ -420,8 +422,6 @@ void diablo_parse_flags(int argc, char **argv)
 		} else if (strcasecmp("-b", argv[i]) == 0) {
 			debug_mode_key_b = 1;
 		*/
-		} else if (strcasecmp("-d", argv[i]) == 0) {
-			debug_mode_key_d = TRUE;
 		} else if (strcasecmp("-i", argv[i]) == 0) {
 			debug_mode_key_i = TRUE;
 			/*
