diff --git a/Source/diablo.cpp b/Source/diablo.cpp
index 2217a444..b5917269 100644
--- a/Source/diablo.cpp
+++ b/Source/diablo.cpp
@@ -117,6 +117,7 @@ static void print_help_and_exit()
 	printf("    %-20s %-30s\n", "-f", "Display frames per second");
 	printf("    %-20s %-30s\n", "-x", "Run in windowed mode");
 	printf("    %-20s %-30s\n", "--spawn", "Force spawn mode even if diabdat.mpq is found");
+	printf("    %-20s %-30s\n", "-d", "Increaased item drops");
 #ifdef HELLFIRE
 	printf("    %-20s %-30s\n", "--theoquest", "Enable the Theo quest");
 	printf("    %-20s %-30s\n", "--cowquest", "Enable the Cow quest");
@@ -164,6 +165,8 @@ static void diablo_parse_flags(int argc, char **argv)
 			fullscreen = FALSE;
 		} else if (strcasecmp("--spawn", argv[i]) == 0) {
 			forceSpawn = TRUE;
+		} else if (strcasecmp("-d", argv[i]) == 0) {
+			debug_mode_key_d = TRUE;
 #ifdef HELLFIRE
 		} else if (strcasecmp("--theoquest", argv[i]) == 0) {
 			UseTheoQuest = TRUE;
